refactor(api): Add missing firmware update api + clean up api so its easier to consume by the robot-server.

vegano1 opened the pull request:
# Overview

The firmware update API was missing a few pieces to work with the firmware update flow from the desktop app, this pull request fixes those issues and exposes the firmware update flags to the instruments so the robot-server is able to get that data easily.

# Test Plan

- [x] Make sure all firmware is updated on startup when `enableOT3FirmwareUpdates` is enabled
- [x] Make sure we can start firmware updates from the robot-server [branch](https://github.com/Opentrons/opentrons/tree/robot-server_add_firmware_update_endpoints) that implements the endpoints.
- [x] Make sure we can monitor firmware update status from the robot-server [branch](https://github.com/Opentrons/opentrons/tree/robot-server_add_firmware_update_endpoints) that implements the endpoints.

# Changelog

- Added OT3Controller.get_instrument_update to check if an instrument needs to be updated
- Changed OT3Controller.get_update_progress so it now just returns set of UpdateStatus which contains OT3SubSystem, state, progress without total progress
- Added InstrumentUpdateStatus to hardware_control/types to be used by robot_server to consume
UpdateProgress class no longer stores total_progress and update, get_progress functions only return Set[UpdateStatus]
- Added update_instrument_firmware just for updating instrument firmware
- Added fw_version and fw_update_required member variables and setters to Gripper and Pipette objects,
Gripper, Pipette fw_version, and fw_update_required are serialized with self.as_dict so they are available in PipetteDict and GripperDict objects.
- Added OT3SubSystem to repl so we can call update_firmware({OT3SubSystem}) from repl to update the firmware
- Added force flag to update_firmware and check_firmware_updates so we can just specify a set of nodes to check with the `nodes` arg instead of that being the `forced` nodes.
- Changed check_firmware_updates so it also returns the next_version and filepath in the format Dict[NodeId, Tuple[int, str]]
- simplified progress reporting so we are only keeping track of individual update progress and not the total progress
- removed enableOT3FirmwareUpdates feature flag

# Review requests

- is there anything this is missing?

# Risk assessment

Low, OT3 dev work

vegano1 committed:
wip

vegano1 committed:
- Added OT3Controller.get_instrument_updates to get which instruments need to be updated
- Changed OT3Controller.get_update_progress so it now just returns set of UpdateStatus which contains OT3SubSystem, state, progress without total progress
- Added InstrumentUpdateStatus to hardware_control/types to be used by robot_server to consume
UpdateProgress class no longer stores total_progress and update, get_progress functions only return Set[UpdateStatus]
- Added OT3api._get_pipette_subtypes to convert attached pipettes to their PipetteSubTypes
- Added OT3api._cache_fw_update_status to refresh the firmware update version and update_required flags in Gripper and Pipette objects, this happens when we cache instruments.
- Renamed  get_update_progress to get_firmware_update_progress and now returns Dict[OT3Mount, - InstrumentUpdateStatus]
- Added update_instrument_firmware just for updating instrument firmware
- update_firmware is now returns  AsyncIterator[Set[UpdateStatus]] and drives the firmware updates
- Added mount_from_subsystem, subsystem_from_mount convertion functions to ot3utils
- Added fw_version and fw_update_required member variables and setters to Gripper and Pipette objects,
Gripper, Pipette fw_version and fw_update_required are serialized with self.as_dict so they are available in PipetteDict and GripperDict objects.
Added OT3SubSystem to repl so we can call update_firmware({OT3SubSystem}) from repl to update the firmware

vegano1 committed:
Merge branch 'edge' into firmware_updates_wip

vegano1 committed:
Merge branch 'firmware_updates_wip' of https://github.com/Opentrons/opentrons into firmware_updates_wip

vegano1 committed:
fix update firmware call

vegano1 committed:
Changes to make robot-server call and usage easier.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -229,11 +223,47 @@ def _attached_pipettes_to_nodes(
+                available_updates[mount] = device_info.version
+
+        # set our internal update required flag
+        self.update_required = bool(available_updates)
```
this probably wants to be `self.update_required |= bool(available_updates)`, since you could have an X controlelr that needs an update that you don't want to get stepped on because your instruments don't need any

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -229,11 +223,47 @@ def _attached_pipettes_to_nodes(
             pipette_nodes[node_id] = pipette_type
         return pipette_nodes
 
+    def get_instrument_updates(
```
looking at the way this is called, it might make it a lot nicer to have this specify just a single pipette at a time: `def get_instrument_update(self, mount: OT3Mount, pipette_type: PipetteSubType)`. This seems like it's removing functionality, but the way that we're calling this is so we can iterate through it; and we're iterating through things here; it's just some extra work that we can fix.

Let's remove logic from here by just making this, "do i have an update for the device on the left mount, which is this kind of device".

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -356,6 +362,95 @@ def fw_version(self) -> str:
+        # fetch firmware update info
+        attached_pipettes = self._get_pipette_subtypes()
+        updates = self._backend.get_instrument_updates(attached_pipettes)
+
```
here, we're implicitly iterating through pipettes inside `get_instrument_updates`, and explicitly iterating on line 396. We can collapse those into just one thing by making it `get_instrument_update(mount, pipette_type)`.

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -356,6 +362,95 @@ def fw_version(self) -> str:
+                pipette_subtypes[mount] = subtypes[pipette.channels]
+        return pipette_subtypes
+
+    def _cache_fw_update_status(self) -> None:
```
This looks like it will work, but it's a very "set attributes of objects" type function where it's kind of hard to predict what it does at a glance. It's accessing other information methods, it's setting attributes on other objects owned through intermediaries...

Let's make this just take a single mount and instrument, or even just a mount, and return the version to update.

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -356,6 +362,95 @@ def fw_version(self) -> str:
+        for mount, pipette in pipettes.items():
+            if pipette:
+                fw_version = updates.get(mount)
+                pipette.fw_version = fw_version or 0
```
This doesn't quite line up to me. `fw_version` just from the name feels like it is "whatever the current firmware on the pipette is", which doesn't really line up with needing an update, right? Here, it appears to be something else since we're doing `fw_update_required = bool(fw_version)`. So what does `fw_version` represent? The version to be updated to?

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -356,6 +362,95 @@ def fw_version(self) -> str:
+            if pipette:
+                fw_version = updates.get(mount)
+                pipette.fw_version = fw_version or 0
+                pipette.fw_update_required = bool(fw_version)
```
If we have logic this simple, and we're already making `fw_version` and `fw_update_required` _properties_, let's just implement `fw_update_required` to do this:
```python
@property
def fw_update_required(self) -> bool:
    return bool(self._fw_version)
```

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -356,6 +362,95 @@ def fw_version(self) -> str:
+        mounts = mounts or set()
+        subsystems = {subsystem_from_mount(mount) for mount in mounts}
+        async for update_status in self.update_firmware(subsystems):
+            mod_log.debug(update_status)
```
no progress for this one?

vegano1 replied:
The specific issue I faced which prompted me to have a function that consumed an iterator but did not return an iterator itself, was calling this from an ayincio.create_task function when we check for updates on startup. Which caused the task to fail to start since it was running under a different asyncio loop context. Maybe because the hardware controller runs as a separate thread?

sfoster1 replied:
yeah you might have to run it under that loop

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -356,6 +362,95 @@ def fw_version(self) -> str:
+        gripper = self._gripper_handler.gripper
+        if gripper:
+            fw_version = updates.get(OT3Mount.GRIPPER)
+            gripper.fw_version = fw_version or 0
```
Also how do these flags get reset after a firmware update?

vegano1 replied:
These flags get set/reset in this function based on whether updates are available or not.

sfoster1 replied:
Can we reset them after a firmware update?

vegano1 replied:
this now gets reset when we check for updates when cache_instruments is called
1. when we finish doing instrument updates
2. when the client queries the current instruments attached

sfoster1 commented on api/src/opentrons/hardware_control/instruments/ot3/pipette_handler.py:
```diff
@@ -204,6 +204,8 @@ def get_attached_instrument(self, mount: OT3Mount) -> PipetteDict:
                 "display_name",
                 "tip_length",
                 "model",
+                "fw_version",
```
is there any way we can possibly avoid this? This is old code that I think we won't be using here

vegano1 replied:
Not at the moment, as far as I understand this mechanism will eventually be separated into state and config objects but the scaffolding for that is not yet setup.

sfoster1 commented on api/src/opentrons/hardware_control/types.py:
```diff
@@ -296,19 +305,33 @@ class UpdateStatus(NamedTuple):
+    progress: int
+
+    @classmethod
+    def update(self, status: UpdateState, progress: int) -> None:
```
This can't be `classmethod` and have `self`, right?

sfoster1 reviewed (changes requested):
I have some inline comments about making functions a little more pure instead of hitting a bunch of object members, but I think in larger part I'd love to see the current firmware version of an instrument, and whether it needs to be updated, become values that are passed into the `__init__` of the objects in `gripper.py` and `pipette.py` and are immutable values. Those objects go away when the pipette goes away (as it's being updated) and then later get recreated after, with the new correct firmware version.

In fact, I'm a little concerned with the  some of these firmware update flags; I don't really get what `fw_version` does since it is so closely linked with `fw_update_needed`; I don't really love those values getting set directly from over in the hardware controller. I think this is almost there though.

vegano1 committed:
various bug fixes and cleaning up

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -229,17 +232,55 @@ def _attached_pipettes_to_nodes(
+    def get_instrument_update(
+        self, mount: OT3Mount, pipette_subtype: Optional[PipetteSubType] = None
+    ) -> InstrumentFWInfo:
+        """Check wether the given instrument requires an update."""
```
```suggestion
        """Check whether the given instrument requires an update."""
```

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -229,17 +232,55 @@ def _attached_pipettes_to_nodes(
+        next_version = update_info[0] if update_info else None
+        # set our global firmware update state if any updates are available
+        self.update_required |= bool(firmware_updates)
+        instrument_fw_info = InstrumentFWInfo(
```
[nit] i generally like just returning the object construction in code like this, since we're really not doing anything `instrument_fw_info`.

vegano1 replied:
oops, left over from debug print

sfoster1 commented on api/src/opentrons/hardware_control/instruments/ot3/gripper_handler.py:
```diff
@@ -54,15 +53,9 @@ def get_gripper(self) -> Gripper:
-            og_gripper.gripper_id,
-        )
-        self._gripper = new_gripper
+        if self._gripper is not None:
```
I don't think this is a good change - the code that's already there looks like it does something else. We shouldn't do this in this pr unless there's a specific functionality reason to.

vegano1 replied:
This was changed to have a single location for creating the gripper object when we cache_instruments. So we can feed in the `fw_update_info` as part of the __init__.

What the code was doing before was recreating the Gripper but keeping all the same configs, which means it was resetting the state which is happening here.

sfoster1 replied:
Okay, that makes sense. However, one thing we've learned through hard experience is that it's very rarely a good idea to reset this stuff by resetting individual class members - it's very easy to forget to reset a new one when you add it, and it requires additional reset paths in the code. That's why the old code here created a whole new Gripper object, and that's the way we should do it here - just make it `new_gripper = Gripper(config, load_...offset(...), ...gripper-id, og_gripper.fw_update_info)`. That way you get to typecheck the call to gripper's `__init__`.

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -327,6 +332,21 @@ def is_simulator(self) -> bool:
+        if a non-fw-update can command is attempted to be sent.
+
+        However, there are other operations as part of system startup (homing, etc)
+        that need to happen without triggering `FirmwareUpdateRequired`. So this
```
but on startup, we immediately _do_ the firmware updates if they're needed, right? before we do stuff like home. is this only a problem when the firmware update feature flag is off?

vegano1 replied:
correct, only when the feature flag is disabled

The reason why this happens is that we get the InstrumentFWInfo when we `cache_instruments` before we create the instrument object. Creating InstrumentFWInfo checks for updates and sets our internal `update_required` flag, which makes sense. If something on the system needs an update, set the internal flag.

So this `initializing` flag lets us set our internal `update_required` flag, while also letting the system go through its normal init procedure.

sfoster1 replied:
I don't think we should add this if it's a problem only when the flag is disabled. I think the better approach is instead to make the `requires_update` decorator check the feature flag or something. This code all becomes unused once we remove the feature flag, right?

vegano1 replied:
agreed

sfoster1 commented on api/src/opentrons/__init__.py:
```diff
@@ -163,7 +163,7 @@ async def _blink() -> None:
         if should_use_ot3() and ff.enable_ot3_firmware_updates():
             log.info("Checking firmware updates")
-            await hardware.start_firmware_updates()
+            await hardware.update_instrument_firmware()
```
This is not a good change to make. This is where we update all the stuff that is _not_ instruments - the x and y and head controllers - _in addition_ to whatever instruments happen to be attached at boot. This may be the reason you had to add the `done_initializing` - because if you're not updating the X controller, then the code will correctly say there are still updates to do and not let you home.

vegano1 replied:
This is just bad naming on my end, `update_instrument_firmware` actually updates everything connected to the system. Will fix that

sfoster1 replied:
As I read it, `update_instrument_firmware` does this:
```python
        subsystems = {subsystem_from_mount(mount)} if mount else set()
        async for update_status in self.update_firmware(subsystems):
            mod_log.debug(update_status)
```

which will specify only the subsystems for the specific mount to `update_firmware`, which is how we derive the `nodes` set, which goes into the update checker and limits what it returns.

If I'm misunderstanding and `update_instrument_firmware` will actually update everything - including the other mounts? - then that seems like wrong behavior for `update_instrument_firmware` vs `update_firmware`, right?

vegano1 replied:
agreed

sfoster1 reviewed (changes requested):
The structure looks good but I'm concerned about two and a half changes:
- I think the initializing thing is not necessary and is there only because of a change we should _not _ make, which is programming just instruments in the module init. That should program everything.
- That change in `gripper_handler` doesn't really look correct to me, I don't think it's reloading things properly? Even if its functionality is the same, it's not necessary to have in this PR.
- That gripper

vegano1 committed:
Update api/src/opentrons/hardware_control/backends/ot3controller.py

Co-authored-by: Seth Foster <seth@opentrons.com>

vegano1 committed:
remove initialized and just use feature_flag to block updates
check that an instrument is attached before attempting to update
check that a node is not already updating before attempting to update
more bug fixes

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -276,14 +317,12 @@ async def update_to_default_current_settings(self, gantry_load: GantryLoad) -> N
         )
         await self.set_default_currents()
 
-    @requires_update
```
want to keep these now that the decorator checks the feature flag right

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -120,6 +121,7 @@ def __init__(
         self._strict_attached = bool(strict_attached_instruments)
         self._stubbed_attached_modules = attached_modules
         self._update_required = False
+        self._initialized = False
```
don't need these anymore right

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -162,6 +164,15 @@ def _sanitize_attached_instrument(
 
+    @property
+    def initialized(self) -> bool:
+        """True when the hardware controller has initialized and is ready."""
```
ditto

sfoster1 reviewed (approved):
Almost there, and approving so as not to block things, but let's just get that gripper reset back to building a new object - we have history here with resetting state on long-lived objects that can be summed up as "don't, just remove and replace the object as a whole", since that way it can get typechecked and it's hard to forget things. Once that's done, this looks great!

vegano1 committed:
revert gripper changes and just pass on old gripper fw_update_info
scplit cache_instruments into _cache_intruments and _configure_instruments so we have more control when initializing the hardware controller.

vegano1 committed:
fix unit tests

vegano1 committed:
- remove ot3EnableFirmwareUpdates feature flag
- move firmware update start from opentrons.__init__ to OT3API so we can manage OT3Controller.initalized better
- remove update_required decorator from internal hardware_controller functions used to get/set info over can
- make sure we update fw_update_info when caching instruments

vegano1 committed:
Merge branch 'edge' into firmware_updates_wip

vegano1 committed:
mypy and fix unit test

sfoster1 reviewed (approved):
Nice, like the changes, looks good to me!

vegano1 committed:
oh mypy

vegano1 committed:
remove unused imports

vegano1 committed:
fix gripperDict issue

vegano1 committed:
fix unit test, dont use typeguard.check_type when comparing PipetteDict as the values have changed

vegano1 committed:
remove unused imports

vegano1 committed:
comment unused var
